src/_posts/2020-07-18-packaging-with-packer.md

```diff
@@ -0,0 +1,603 @@
+            ]
+        },
+        {
+            "type": "shell",
+            "inline": [
+                "cd /tmp",
+                "curl -L -O curl -L -O https://artifacts.elastic.co/downloads/beats/metricbeat/metricbeat-7.7.0-amd64.deb",
+                "sudo dpkg -i metricbeat-7.7.0-amd64.deb",
+                "rm metricbeat-7.7.0-amd64.deb"
+            ]
+        }
+    ]
+}
+```
+
+With the base image ready we can change the current file to use the base image, we change the `source_ami_filter` to pick the latest `base-image-example-*` (don't forget to update the `owners` field). Then remove the programs that were previously installed.
+
+```json
+{
+    "variables": {
+        "aws_access_key": "",
+        "aws_secret_key": "",
+        "docker_hub_password": ""
+    },
+    "sensitive-variables": [
+        "docker_hub_password"
+    ],
+    "builders": [
+        {
+            "type": "amazon-ebs",
+            "access_key": "{{user `aws_access_key`}}",
+            "secret_key": "{{user `aws_secret_key`}}",
+            "region": "eu-west-2",
+            "source_ami_filter": {
+                "filters": {
+                    "virtualization-type": "hvm",
+                    "name": "base-image-example-*",
+                    "root-device-type": "ebs"
+                },
+                "owners": [
+                    "<owner_id>"
+                ],
+                "most_recent": true
+            },
+            "instance_type": "t2.micro",
+            "ssh_username": "ubuntu",
+            "ami_name": "packer-example-{{timestamp}}"
+        },
+        {
+            "type": "docker",
+            "image": "ubuntu:18.04",
+            "commit": true,
+            "changes": [
+                "LABEL mantainer.name='Andre Torres'",
+                "EXPOSE 8080"
+            ]
+        }
+    ],
+    "provisioners": [
+        {
+            "type": "shell",
+            "inline": [
+                "sleep 30"
+            ]
+        },
+        {
+            "type": "shell",
+            "script": "./fix-sudo.sh"
+        },
+        {
+            "type": "shell",
+            "inline": [
+                "sudo apt-get update -y",
+                "sudo apt-get install -y openjdk-11-jre-headless"
+            ]
+        }
+    ],
+    "post-processors": [
+        [
+            {
+                "type": "docker-tag",
+                "repository": "andretorrescodurance/packer_test",
+                "tag": [
+                    "0.1"
+                ],
+                "only": [
+                    "docker"
+                ]
+            },
+            {
+                "type": "docker-push",
+                "login": "true",
+                "login_username": "andretorrescodurance",
+                "login_password": "{{user `docker_hub_password`}}",
+                "only": [
+                    "docker"
+                ]
+            }
+        ]
+    ]
+}
+```
+
+## Wrapping up
```

Comment:
I would add two additional points:

* Mention that you can use tools like Puppet to do the provisioning. This allows someone who already uses these tools to create AMIs from the same codebase they already have.

* Talk about how you can add Packer into CI/CD flows to easily create and manage immutable infrastructure. A code change leads to a Packer run which can then update, say, an ASG `LaunchConfiguration`.

Reply:
The part about being able to use Puppet/Chef I'm thinking I could add together with the provisioners inline 100.